src/profiler.ts

```diff
@@ -13,57 +13,292 @@
+  const pwriter = perftools.profiles.Profile.encode(p);
+  const buffer = new Buffer(pwriter.finish());
+  const gzBuf = await gzip(buffer);
+  return gzBuf.toString('base64');
+}
+
+/**
+ * Polls Stackdriver Profiler server for instructions on behalf of a task and
+ * collects and uploads profiles as requested
+ */
+export class Profiler extends common.ServiceObject {
+  private config: ProfilerConfig;
+  private logger: Logger;
+  private profileTypes: string[];
+
+  // Public for testing.
+  timeProfiler: TimeProfiler|undefined;
 
   constructor(config: ProfilerConfig) {
-    this.config = config;
+    config = common.util.normalizeArguments(null, config);
+    const serviceConfig = {
+      baseUrl: API,
+      scopes: ['https://www.googleapis.com/auth/monitoring.write'],
+      packageJson: pjson,
+    };
+    super({parent: new common.Service(serviceConfig, config), baseUrl: '/'});
 
-    this.service = new common.Service(
-        {
-          baseUrl: 'https://cloudprofiler.googleapis.com/v2',
-          scopes: ['https://www.googleapis.com/auth/monitoring.write'],
-        },
-        this.config);
+    this.config = config;
 
     this.logger = new common.logger({
       level: common.logger.LEVELS[config.logLevel as number],
       tag: pjson.name
     });
+
+    // TODO: enable heap profiling once heap-profiler implemented.
+    this.profileTypes = [];
+    if (!this.config.disableTime) {
+      this.profileTypes.push(ProfileTypes.Wall);
+      this.timeProfiler =
+          new TimeProfiler(this.config.timeSamplingIntervalMicros);
+    }
   }
 
-  // Begins collection and uploading of profiles.
-  // If profiling fails or another problem is encountered, the returned promise
-  // will be rejected.
-  // TODO: implement
-  // TODO: explain failure situations.
+  /**
+   * Starts and endless loop to poll Stackdriver Profiler server for
+   * instructions, and collects and uploads profiles as requested.
+   * If there is a problem when collecting a profile or uploading a profile to
+   * Stackdriver Profiler, this problem will be logged at the debug level.
+   * If there is a problem polling Stackdriver Profiler for instructions
+   * on the type of profile created, an error will be thrown.
+   */
   async start(): Promise<void> {
-    return Promise.reject(
-        new Error('start() is unimplemented for ProfileAgent'));
+    return this.pollProfilerService();
+  }
+
+
+  /**
+   * Endlessly polls the profiler server for instructions, and collects and
+   * uploads profiles as requested.
+   *
+   * The profiler server will be polled for instructions at most once every
+   * minTimeBetweenProfilesMillis.
+   */
+  async pollProfilerService(): Promise<void> {
+    const startCreateMillis = Date.now();
+    const prof = await this.createProfile();
+    await this.profileAndUpload(prof);
+    const endCreateMillis = Date.now();
+
+    // Schedule the next profile.
+    const delayMillis = Math.max(
```

Comment:
We shouldn't have this delay. The profiles are collected at exponential distribution points with the mean of 1 minute, capping the min will skew the mean.

Reply:
The concern I have here is that a misbehaving server, proxy or network can cause `createProfile` to return early, causing the agent to spin the CPU on the node process. Given that we're talking about a single thread here, this can have a pretty major impact on the user's application.

We need to defend against this possibility. Is there no way to avoid the skew?

Reply:
Errors like that should result in an HTTP / response error which are handled and back off. It might make sense to validate that the response from the server has the unique profile ID set.

Reply:
I'm okay with any solution that guarantees that we won't spin the CPU.